### PR TITLE
bugfix(react-utilities): omit IntrinsicElement props from ComponentProps that would collide with slot names

### PR DESCRIPTION
**Pull request by bsunderhus**
## Previous Behavior

As https://github.com/microsoft/fluentui/issues/29596 describes, some properties that were declared by `@types/react` were clashing with slot names, most common case would be the one involving the `content` slot.

## New Behavior

1. omits properties declared by `@types/react` in favor of the slot names provided

## Related Issue(s)

- Fixes https://github.com/microsoft/fluentui/issues/29596

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/Avatar-mount_base.html">659</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/Avatar-mount.html">626</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/Button-mount_base.html">321</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/Button-mount.html">312</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/Field-mount_base.html">1113</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/Field-mount.html">1144</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/FluentProvider-mount_base.html">692</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/FluentProvider-mount.html">721</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/FluentProviderWithTheme-mount_base.html">81</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/FluentProviderWithTheme-mount.html">92</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">61</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">66</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">76</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">73</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/MakeStyles-mount_base.html">864</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/MakeStyles-mount.html">858</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/Persona-mount_base.html">1732</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/Persona-mount.html">1706</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/SpinButton-mount_base.html">1393</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29865/perf-test-react-components/SpinButton-mount.html">1386</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                   |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>global-context</samp> <br /> <abbr title='bundle-size/createContext.fixture.js'>createContext</abbr>                          |          `510 B`<br />`330 B` |
| <samp>global-context</samp> <br /> <abbr title='bundle-size/createContextSelector.fixture.js'>createContextSelector</abbr>          |          `537 B`<br />`342 B` |
| <samp>react-accordion</samp> <br /> <abbr title='bundle-size/Accordion.fixture.js'>Accordion (including children components)</abbr> |   `93.44 kB`<br />`28.513 kB` |
| <samp>react-badge</samp> <br /> <abbr title='bundle-size/Badge.fixture.js'>Badge</abbr>                                             |   `27.264 kB`<br />`8.808 kB` |
| <samp>react-badge</samp> <br /> <abbr title='bundle-size/CounterBadge.fixture.js'>CounterBadge</abbr>                               |   `28.165 kB`<br />`9.101 kB` |
| <samp>react-badge</samp> <br /> <abbr title='bundle-size/PresenseBadge.fixture.js'>PresenceBadge</abbr>                             |    `25.67 kB`<br />`9.313 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/Button.fixture.js'>Button</abbr>                                          |  `40.074 kB`<br />`11.292 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/CompoundButton.fixture.js'>CompoundButton</abbr>                          |   `47.43 kB`<br />`12.787 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/MenuButton.fixture.js'>MenuButton</abbr>                                  |  `44.848 kB`<br />`12.668 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/SplitButton.fixture.js'>SplitButton</abbr>                                |  `52.862 kB`<br />`14.254 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/ToggleButton.fixture.js'>ToggleButton</abbr>                              |  `57.122 kB`<br />`13.194 kB` |
| <samp>react-calendar-compat</samp> <br /> <abbr title='bundle-size/Calendar.fixture.js'>Calendar Compat</abbr>                      | `141.822 kB`<br />`36.959 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/Card.All.fixture.js'>Card - All</abbr>                                      |  `93.423 kB`<br />`27.059 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/Card.fixture.js'>Card</abbr>                                                |  `88.247 kB`<br />`25.559 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardFooter.fixture.js'>CardFooter</abbr>                                    |   `13.423 kB`<br />`5.481 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardHeader.fixture.js'>CardHeader</abbr>                                    |   `15.666 kB`<br />`6.235 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardPreview.fixture.js'>CardPreview</abbr>                                  |   `14.374 kB`<br />`5.854 kB` |
| <samp>react-checkbox</samp> <br /> <abbr title='bundle-size/Checkbox.fixture.js'>Checkbox</abbr>                                    |  `34.424 kB`<br />`11.829 kB` |
| <samp>react-dialog</samp> <br /> <abbr title='bundle-size/Dialog.fixture.js'>Dialog (including children components)</abbr>          |  `95.021 kB`<br />`28.381 kB` |
| <samp>react-divider</samp> <br /> <abbr title='bundle-size/Divider.fixture.js'>Divider</abbr>                                       |   `21.179 kB`<br />`7.839 kB` |
| <samp>react-field</samp> <br /> <abbr title='bundle-size/Field.fixture.js'>Field</abbr>                                             |   `22.788 kB`<br />`8.639 kB` |
| <samp>react-image</samp> <br /> <abbr title='bundle-size/Image.fixture.js'>Image</abbr>                                             |   `16.096 kB`<br />`6.317 kB` |
| <samp>react-input</samp> <br /> <abbr title='bundle-size/Input.fixture.js'>Input</abbr>                                             |   `27.247 kB`<br />`9.194 kB` |
| <samp>react-jsx-runtime</samp> <br /> <abbr title='bundle-size/Classic.fixture.js'>Classic Pragma</abbr>                            |       `1.049 kB`<br />`531 B` |
| <samp>react-jsx-runtime</samp> <br /> <abbr title='bundle-size/DevRuntime.fixture.js'>JSX Dev Runtime</abbr>                        |     `4.14 kB`<br />`1.741 kB` |
| <samp>react-jsx-runtime</samp> <br /> <abbr title='bundle-size/Runtime.fixture.js'>JSX Runtime</abbr>                               |    `4.734 kB`<br />`1.983 kB` |
| <samp>react-label</samp> <br /> <abbr title='bundle-size/Label.fixture.js'>Label</abbr>                                             |   `14.508 kB`<br />`5.857 kB` |
| <samp>react-link</samp> <br /> <abbr title='bundle-size/Label.fixture.js'>Link</abbr>                                               |     `17.4 kB`<br />`6.977 kB` |
| <samp>react-menu</samp> <br /> <abbr title='bundle-size/Menu.fixture.js'>Menu (including children components)</abbr>                | `140.971 kB`<br />`43.239 kB` |
| <samp>react-menu</samp> <br /> <abbr title='bundle-size/Menu.Selectable.fixture.js'>Menu (including selectable components)</abbr>   | `143.657 kB`<br />`43.747 kB` |
| <samp>react-message-bar</samp> <br /> <abbr title='bundle-size/Default.fixture.js'>MessageBar (all components)</abbr>               |    `25.54 kB`<br />`9.102 kB` |
| <samp>react-overflow</samp> <br /> <abbr title='bundle-size/hooks.fixture.js'>hooks only</abbr>                                     |   `12.837 kB`<br />`4.837 kB` |
| <samp>react-popover</samp> <br /> <abbr title='bundle-size/Popover.fixture.js'>Popover</abbr>                                       | `120.349 kB`<br />`37.903 kB` |
| <samp>react-portal</samp> <br /> <abbr title='bundle-size/Portal.fixture.js'>Portal</abbr>                                          |   `12.571 kB`<br />`4.594 kB` |
| <samp>react-positioning</samp> <br /> <abbr title='bundle-size/usePopper.fixture.js'>usePositioning</abbr>                          |   `25.637 kB`<br />`9.263 kB` |
| <samp>react-progress</samp> <br /> <abbr title='bundle-size/ProgressBar.fixture.js'>ProgressBar</abbr>                              |   `17.788 kB`<br />`6.986 kB` |
| <samp>react-provider</samp> <br /> <abbr title='bundle-size/FluentProvider.fixture.js'>FluentProvider</abbr>                        |    `23.008 kB`<br />`8.48 kB` |
| <samp>react-radio</samp> <br /> <abbr title='bundle-size/Radio.fixture.js'>Radio</abbr>                                             |  `31.437 kB`<br />`10.314 kB` |
| <samp>react-radio</samp> <br /> <abbr title='bundle-size/RadioGroup.fixture.js'>RadioGroup</abbr>                                   |    `15.704 kB`<br />`6.35 kB` |
| <samp>react-select</samp> <br /> <abbr title='bundle-size/Select.fixture.js'>Select</abbr>                                          |  `28.968 kB`<br />`10.269 kB` |
| <samp>react-slider</samp> <br /> <abbr title='bundle-size/Slider.fixture.js'>Slider</abbr>                                          |   `38.52 kB`<br />`12.646 kB` |
| <samp>react-spinbutton</samp> <br /> <abbr title='bundle-size/SpinButton.fixture.js'>SpinButton</abbr>                              |  `37.133 kB`<br />`11.856 kB` |
| <samp>react-spinner</samp> <br /> <abbr title='bundle-size/Spinner.fixture.js'>Spinner</abbr>                                       |   `23.762 kB`<br />`8.556 kB` |
| <samp>react-switch</samp> <br /> <abbr title='bundle-size/Switch.fixture.js'>Switch</abbr>                                          |  `33.902 kB`<br />`10.944 kB` |
| <samp>react-text</samp> <br /> <abbr title='bundle-size/Text.Default.fixture.js'>Text - Default</abbr>                              |   `17.065 kB`<br />`6.645 kB` |
| <samp>react-text</samp> <br /> <abbr title='bundle-size/Text.Wrappers.fixture.js'>Text - Wrappers</abbr>                            |   `20.237 kB`<br />`6.965 kB` |
| <samp>react-textarea</samp> <br /> <abbr title='bundle-size/Textarea.fixture.js'>Textarea</abbr>                                    |  `31.306 kB`<br />`10.549 kB` |
| <samp>react-toast</samp> <br /> <abbr title='bundle-size/Toast.fixture.js'>Toast (including Toaster)</abbr>                         |  `93.195 kB`<br />`27.948 kB` |
| <samp>react-tooltip</samp> <br /> <abbr title='bundle-size/Tooltip.fixture.js'>Tooltip</abbr>                                       |  `53.173 kB`<br />`18.712 kB` |
| <samp>react-utilities</samp> <br /> <abbr title='bundle-size/SSRProvider.fixture.js'>SSRProvider</abbr>                             |          `180 B`<br />`159 B` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/abb2dc5195c3e36c0f80e1a7fd4a7ebbdcb01f10'>abb2dc5195c3e36c0f80e1a7fd4a7ebbdcb01f10</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Review by miroslavstastny — APPROVED** [review]

File: packages/react-components/react-utilities/src/compose/types.ts [hunk 1]
```diff
@@ -169,15 +169,24 @@ export type ExtractSlotProps<S> = Exclude<S, SlotShorthandValue | null | undefin
 export type ComponentProps<Slots extends SlotPropsRecord, Primary extends keyof Slots = 'root'> =
   // Include a prop for each slot (see note below about the Omit)
   Omit<Slots, Primary & 'root'> &
-    // Include all of the props of the primary slot inline in the component's props
-    PropsWithoutRef<ExtractSlotProps<Slots[Primary]>>;
+    // Include all of the props of the primary slot inline in the component's props,
+    // omits the props that would collide with slots, e.g: content
+    // excluding the Primary property itself, in case root is another component, e.g: DatePick
+    OmitWithoutExpanding<PropsWithoutRef<ExtractSlotProps<Slots[Primary]>>, Exclude<keyof Slots, Primary>>;
```
[the remark this answers is not in the document]
Comment (layershifter) [reply packages/react-components/react-utilities/src/compose/types.ts:175]: > Since this only affects a small number of components, it would be ideal if we could avoid adding complexity to the core ComponentProps type. Also, whenever we add non-configurable/"magic" behavior to these types, they become harder to understand (they're already pretty bad in this respect).
>
> Could we instead create a solution that lets components omit specific intrinsic props when they conflict with slot names?

While I completely agree about the complexity of types, I am uncertain whether the approach involving specific props is a viable solution 😶

- We do not maintain `@types/react`, and these types do not adhere to semver ⬇️
- There are no assurances that new properties will be added/removed

In summary, if we manage specific properties without a guard, we risk encountering issues as seen in #29596.  IMO, this is a significant drawback as it does not ensure a future-proof solution.

[the remark this answers is not in the document]
Comment (layershifter) [reply packages/react-components/react-utilities/src/compose/types.ts:175]: P.S. If I need to pick an option, I would pick 1 as it adds nothing new.

[the remark this answers is not in the document]
Comment (bsunderhus) [reply packages/react-components/react-utilities/src/compose/types.ts:175]: I guess I wasn't clear enough in what I meant, let me rephrase it: All options provided here  (Option 1 to Option 3) will have the same [Distributive Conditional type](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types) issue, as all of them rely on us externalizing `Omit` to the component development!

Here's a [playground](https://www.typescriptlang.org/play?ts=4.4.4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFCiSxwDeAwiZAHZYcwAKxYVABo4AZQA2EGAF84REuQACBcQFduMVcAD0lPDAC0udhC49UZANw0aMAJ5gsYkljbhTGiVPQBeZjThAuGIpAC4xSRgAHjIAE2AANzIRMgAjVRgYUzIAPgCg4x4NcK9ouMTcmmkbLAAPBnh7R2cQVxMzPgFfOHyAeRBgaLdODX4IQSjRF2GPHlLUHJTCmA1cuAAyHsDeYFwAa0np9s9I4XJl1ZzrGkLUeCnWmY64PwAKMC6So-cOscEASheOTgHFU4nENlu8FqLzgh0exx4+UCaB8ACJ0plTGjkXBtNo4E0nMB0BwpCg4AlkOJgLE4B9xlhYHYUOIYAALCCqADm7LggzI6GQcExWQ4uKJ6NF2Nx+LgWAS-KFHDsuNMLBp+x8TFeCsBPmBtwg4iwADpJNzdQl-tUgniCRceHAAO7QPboAjALiEzk8vmhf64x0wbVReIJbQ5aqRoA) of the Option 1 breaking

Here's a [playground](https://www.typescriptlang.org/play?ts=4.4.4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFCiSxwDeAygDYQwAKxYq2uaABMANHHacAvnCIlyAAQJsArlgB2MZcAD0lPDAC0g8BDXqYqMgG46JxkwCiADxhR9E7r1TTZpMopVzLV0qQ2UYYDZgCKxLGhptBAQaREwsEAgAN1i4GAALLHJKAjI4MF4ZYlJ8woBzYGy1OB4IPlyATzAsMTYcTOA1WrhlNWBTdAGYfTgAClRlXDyUdBq4IWBUXChQAeQYLCFh0dM4LZx9w4AjdpSkAEkNbbVUYFwPFr4ASgA6OAAVPIbFBsDgAd3QD1cAxebw4nlaKwgcEuhWUqAOcFB0SW2H03wI0FByCgQmwBG+t1u-wKp1M6wipmQbDgACIsC51EJ0CMANZqCCgtQs2ZMontFZQVSfODssAcaLoZBwACOWigPLgA1yNLF-06WBYW2AYHgeWQaiEvXQggt0TGaiZVJg+tQYhgSPKWEa8BG9vQvmRWAGQ1l5qEBwpiG0NFl0Hgzq6zS8AHVsRBwmSADxcAB8cAAvEUsCUZRyLegeVh2hACM04AB+WZcUv7ctHPkCpqNgDyIGi2bEFGLZDzAC5mtLx1wbPFY4wE4UAMIkSBmDTvLyZjzodmtrniOEffhYQQksQ8UDE9otzkVqs1g+cdCFigQTgjgspOC9-vbwfEd84AAMiTS8oHaHMZzneN9TgZcTDXGBISeGEjy3OEdzLfcNwRARhHPbYQCvG820rata23AsijfGAP3zL8j1TfJ0xgLNnFcdxD03bcAG0LyI8CAF0c0g2cnAYGDE3g1dzDQyjd1vR94T4PCz1AgTrwU0j7wojCqNfd880LL9pNMcwcL4dCnwIsCIOAuCVzMjRkOhV45IwmyNNEmgF3EEgsFMxDKMLJgvzgOAAJgccPEzMh1kyMhB0ucJ3TUEcwtpDRzHraK4Vi+L0skMSJI6RMWH8wLZK8KiTMcoKuIRLcKrq8yMLzECvx-GBM0q5zHlc3A5OahDWqfHNBxtVtaO8m1UHgcqQAClqNComZygRaLhpkjQj2lfM8zUZQQXiWb4CcKimsW3qYAytB8xZZKYFSlkMu0bRSsKIF+XgJVMiZYBDnWrpYGvJlmOUWolmiMhFWRFKTkyNhkABspeCwEGMoXe7Hue173q9TVFTUG5wvC0xFyiXAeXzJgZi9Pa81miBem+DhajpzJPiK7QcyAA) of the Option 2 breaking

Here's a [playground](https://www.typescriptlang.org/play?ts=4.4.4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFCiSxwDeAwiZAHZYcwAKxYVABo4AZQA2EGCIlTsHACZYoAMQCuHfMAgcAvnCIlyAAQLi13GGuAB6SnhgBaXOx2XUZANw0aNhAhpETCwQCAA3LHQYAAssckoCMjgwAQNiUhi4gHNgCI44fghBOBgATzAsEXEcMOAOLLgNbQ50OpgHOAAKVDVcaJQo2LgFYFRcKFA65BgsBUaOZrhxnBm5gCNSwKQASR4JluBcWT4BVABKADo4ABVo0ZRxSQB3dF2YfdRD48LimAg4NZxNSoWZwJ7AGLBBwXAjQJ7IKAKbAEC5bLY3IYuRQQ5rIcRwABEWAAHjNFOgNABrDgQJ4cAldPHw0pRKAWM5wElgSQQ9DIOAAR2sUEpcDqJSGzJu5SwonGwDA8GiyEU1XQWJGMFx4nRZQqwhK-xSWDy8CaOnQhlIgLqDS5KqUClRiBsNC50HgeriP1QAHUIdEIGoYCw7uIFJQOAAeXgAPjgAF5yH1gOHI0kSWSFOhKVhShACAU4AB+Lq8Tmk7jZ+bU2n5UsAeRAEJjIjIKbT3DI8YAXAUOX3eN5fP50QAJa4AWQAMpzqiBLCUZYNpihKHAafA8c9QX84MqIkswxHuM6AkFkcpuLgsH3ojAYIIezYbEoIpIKlALqEAF6p8TIBc0BZDY3COAAqqIr4QLgqA2AA4pIqCoAipQ2AAouAZQAPpYPOlhbK6XpwJhiqlG8HyHOh+E8OgCaBHAAA+5AIjgtBwEx5BrGgWDsZxZBrFAfHMe2EDiMJ5AhICCgSWQ0RCQxImgFksl1GAwayeIdSUrJC7tLJYAIsgICyagQZQDesnvHgOmKeQTyCV4Ph+OeSBjnhn5LhUBjQMwxhaRworHLoVyThAIwEMAkRwAABgAUqIAAaFwUXUny4NRITuAA2tcMoALoxT26JINgoQROg1RZHgpRwKg7y2nACRnkEAAiozbrS6rHpGPlQJyWG1e0WSVcAubkMpZBnkRMpwKlBwZTRJxFKgUZ5d5mZVjmeYFnACXJfN6WZQutHxkm61xJt5IkYNh1UUtqAMaWPr+jEQYhj13BRtg0IvQGyJRvtKV7Gl91ZbRuUFbGsYMX2P34Bcf0xADQN3Yt4MwKgkMVPlsbDi544ecoXlxLC-VMP52liJIMAhTcyC5oMcQpCtYJxPCPCGnVahgAwhC+fyqA0ygigoAo1aZCgayphCtVk7FHYnhwMW6v8gIoE1VbEwQGhaDoEqru0jMShAILJKc000MRr3RMc8hKKoutajoManGdBSnHAABkzAMYrkbFn2nQ+hWWboEwR6pkrgc1jSdJwPoz2nNl7afRwZD5XAfZcBEUAcsxdta47mjO9GPp4zQujDu6jDEWj3ynGts1XdWub5oWqMgwtx3uGdDEcRdodbSLmwcWPY+lhHaAx4P+g+zbheKMo6gl80UZoz3PA+k3FTQ-349ZxuJrKD4xGiCQWBsOAbg8McdG++PxBSH2xxRk2Lb1zT29kCMYRJCJaxgx-HTrGNsWIyQwG7DDce4DLAvxplGH+uRuyVx8DXT0s1z4LivpwSwIckw4JvstQQUYsGX1cFwW+NNUAwxoFieqYgL6EMofAJMnQWZPkYdgiheDTgcgTPGDgahHg+HofAYkiY4CkKYTwng+80AJgJIAh8OgCT7xfCTMU6BNwazCHiYAcwOHKDKA8N6agsj9AhGQPkAIgE6H3l6RRyjgFqPHhok0WiR77x0CwLSuBKQJiYJ0E0-D4z0LElgC4kgsjBLCGcXQ6ibBLB0BAsE0BKSWjqHEMxFis5nH3rAnggSoy-xsLGBJZSgA) of the Option 3 breaking

The problem here is with externalizing the possibility of modifying unions. When we do it internally we can ensure that the distributive conditional types will be properly applied to ensure unions will not be expanded. Any sort of modification in an union type will have this problem! And since we do have unions on the root type of some of our components, I'm not confident that having this is a solid solution long term.

I must say it again though, although this breaks for unions cases, the components with the conflicting types **DO NOT** have unions in their base type, so this will not be a problem **AT THE MOMENT**

[the remark this answers is not in the document]
Comment (bsunderhus) [reply packages/react-components/react-utilities/src/compose/types.ts:175]: As this is quite explosive, and it can break at any time due to the [Distributive conditional type](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types) I propose we keep it as it is in the PR implementation right now, let's keep any modification of a possible union internal to our type system, to ensure we're properly treating this edge case.

**Review by layershifter — APPROVED** [review]

File: packages/react-components/react-utilities/src/compose/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (behowell) [reply packages/react-components/react-utilities/src/compose/types.ts:175]: Apologies for the delay, this fell off my radar.

I still think we should go with Option 1, which empowers component types to do the right thing. A component can just as easily use `OmitWithoutExpanding` instead of `Omit` in cases where it matters. And unless I'm misunderstanding, it's a bugfix to replace `Omit` with `OmitWithoutExpanding` if the component author forgot, right? (It's not a breaking change).

Regarding @layershifter's comment about `@types/react` potentially breaking in the future... That may be an issue, but it seems like this PR would only help in a very narrow set of circumstances, and there would probably be other breaking changes as well. I don't think we should design around hypothetical future breaks from `@types/react`.

[the remark this answers is not in the document]
Comment (bsunderhus) [reply packages/react-components/react-utilities/src/compose/types.ts]: Ok, I'll converted this PR to go with Option 1

**Review by behowell — APPROVED** [review]

**Comment by bsunderhus** [comment]
Seems like this solution is biting us again while trying to upgrade to v18. 
image
